packages/next/server/config.ts

```diff
@@ -138,7 +140,29 @@ function assignDefaults(userConfig: { [key: string]: any }) {
     {}
   )
 
-  const result = { ...defaultConfig, ...config }
+  // Copy properties we intend to mutate.
+  const copy = (nextConfig: NextConfig): NextConfig => {
+    return {
+      ...nextConfig,
+      ...(nextConfig.pageExtensions && {
+        pageExtensions: [...nextConfig.pageExtensions],
+      }),
+      images: {
+        ...nextConfig.images,
+      },
+      amp: {
+        ...nextConfig.amp,
+      },
+      httpAgentOptions: {
+        ...nextConfig.httpAgentOptions,
+      },
+      experimental: {
+        ...nextConfig.experimental,
+      },
+    }
+  }
+
+  const result = { ...copy(defaultConfig), ...config }
```

Comment:
You're cloning twice
1.  copy() performs a clone
2. `...`  performs a shallow clone

Reply:
Yup! We could do `Object.asign` or similiar, but you still need to merge the `defaultConfig` copy with the `config` object. Feel free to make a suggestion on the merge approach, which right now is the 2nd shallow clone.